adapters/repos/db/vector/hnsw/index.go

```diff
@@ -711,3 +684,86 @@ func IsHNSWIndex(index any) bool {
 	_, ok := index.(*hnsw)
 	return ok
 }
+
+type pair struct {
+	nodeId uint64
+	level  int
+}
+
+func (h *hnsw) calculateUnreachablePoints() []uint64 {
+	h.RLock()
+	defer h.RUnlock()
+
+	visitedPairs := make(map[pair]bool)
+	candidateList := []pair{{h.entryPointID, h.currentMaximumLayer}}
+
+	for len(candidateList) > 0 {
+		currentNode := candidateList[len(candidateList)-1]
+		candidateList = candidateList[:len(candidateList)-1]
+		if !visitedPairs[currentNode] {
+			visitedPairs[currentNode] = true
+			neighbors := h.nodes[currentNode.nodeId].connectionsAtLowerLevelsNoLock(currentNode.level, visitedPairs)
+			candidateList = append(candidateList, neighbors...)
+		}
+	}
+
+	visitedNodes := make(map[uint64]bool, len(visitedPairs))
+	for k, v := range visitedPairs {
+		if v {
+			visitedNodes[k.nodeId] = true
+		}
+	}
+
+	unvisitedNodes := []uint64{}
+	for i := 0; i < len(h.nodes); i++ {
+		if h.nodes[i] == nil {
```

Comment:
There is a missing lock here. Basically, `h.RLock` is there to lock the `h.nodes` slice pointer, to prevent concurrent full replacement (i.e `h.nodes = someOtherSlice`).
But each individual slot of the slice can still be accessed and modified concurrently. To prevent that, you need to use the shardedLocks.

```go
var id uint64
h.shardedNodeLocks.RLock(i)
if h.nodes[i] != nil {
  id = h.nodes[i].id
}
h.shardedNodeLocks.RUnlock(id)
if id == 0 {
  continue
}
```

Reply:
Should `RUnlock(i)` be done after line 724?

Reply:
No need if you have extracted the id already:

```go
if !visitedNodes[uint64(i)] {
	unvisitedNodes = append(unvisitedNodes, id)
}
```